File: benchmarks/bench_sort.js
#!/usr/bin/env node
// Node counterpart of bench_sort.strada (identical workload).
let seed = 42;
function lcg() { seed = (seed * 1103515245 + 12345) % 2147483648; return seed; }

const ints = []; for (let i = 0; i < 1000000; i++) ints.push(lcg() % 10000000);
const strs = []; for (let i = 0; i < 500000; i++) strs.push("key_" + (lcg() % 1000000) + "_suffix");

let t0 = Date.now();
const si = [...ints].sort((a, b) => a - b);
let t1 = Date.now();
console.log("int-sort:", si[0], (t1 - t0) / 1000);

const ss = [...strs].sort();
let t2 = Date.now();
console.log("str-sort:", ss[0].length, (t2 - t1) / 1000);

const half = ints.slice(0, 500000);
const sc = half.sort((a, b) => b - a);
let t3 = Date.now();
console.log("cmp-sort:", sc[0], (t3 - t2) / 1000);

const h = {}; for (let i = 0; i < 200000; i++) h["k" + i] = i;
let hsum = 0;
for (let r = 0; r < 5; r++) { const hk = Object.keys(h).sort(); hsum += hk[0].length; }
let t4 = Date.now();
console.log("hash-sort:", hsum, (t4 - t3) / 1000);
console.log("total:", (t4 - t0) / 1000);

File: benchmarks/bench_sort.pl
#!/usr/bin/env perl
# Perl counterpart of bench_sort.strada (identical workload).
use strict; use warnings; use Time::HiRes qw(time);

my $seed = 42;
sub lcg { $seed = ($seed * 1103515245 + 12345) % 2147483648; $seed }

my @ints; push @ints, lcg() % 10000000 for 1..1000000;
my @strs; push @strs, "key_" . (lcg() % 1000000) . "_suffix" for 1..500000;

my $t0 = time;
my @si = sort { $a <=> $b } @ints;
my $t1 = time;
printf "int-sort: %s %.6f\n", $si[0], $t1 - $t0;

my @ss = sort @strs;
my $t2 = time;
printf "str-sort: %s %.6f\n", length($ss[0]), $t2 - $t1;

my @half = @ints[0..499999];
my @sc = sort { $b <=> $a } @half;
my $t3 = time;
printf "cmp-sort: %s %.6f\n", $sc[0], $t3 - $t2;

my %h; $h{"k$_"} = $_ for 0..199999;
my $hsum = 0;
for (1..5) { my @hk = sort keys %h; $hsum += length($hk[0]); }
my $t4 = time;
printf "hash-sort: %s %.6f\n", $hsum, $t4 - $t3;
printf "total: %.6f\n", $t4 - $t0;
